File: app/src/main/java/at/tomtasche/reader/ui/FindActionModeCallback.java
/*
 * Copyright (C) 2010 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package at.tomtasche.reader.ui;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Point;
import android.graphics.Rect;
import android.text.Editable;
import android.text.Selection;
import android.text.Spannable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.webkit.WebView;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.view.ActionMode;
import at.tomtasche.reader.R;

// https://github.com/aosp-mirror/platform_frameworks_base/blob/a8b1b1a2e62bcca18f52ed31549c93d43728d152/core/java/android/webkit/FindActionModeCallback.java
public class FindActionModeCallback implements ActionMode.Callback, TextWatcher,
        View.OnClickListener, WebView.FindListener {
    private final View mCustomView;
    private final EditText mEditText;
    private final TextView mMatches;
    private WebView mWebView;
    private final InputMethodManager mInput;
    private final Resources mResources;
    private boolean mMatchesFound;
    private int mNumberOfMatches;
    private int mActiveMatchIndex;
    private ActionMode mActionMode;

    public FindActionModeCallback(Context context) {
        mCustomView = LayoutInflater.from(context).inflate(
                R.layout.webview_find, null);
        mEditText = mCustomView.findViewById(
                R.id.edit);
        mEditText.setOnClickListener(this);
        setText("");
        mMatches = mCustomView.findViewById(
                R.id.matches);
        mInput = context.getSystemService(InputMethodManager.class);
        mResources = context.getResources();
    }

    public void finish() {
        mActionMode.finish();
    }

    /**
     * Place text in the text field so it can be searched for.  Need to press
     * the find next or find previous button to find all of the matches.
     */
    public void setText(String text) {
        mEditText.setText(text);
        Spannable span = mEditText.getText();
        int length = span.length();
        // Ideally, we would like to set the selection to the whole field,
        // but this brings up the Text selection CAB, which dismisses this
        // one.
        Selection.setSelection(span, length, length);
        // Necessary each time we set the text, so that this will watch
        // changes to it.
        span.setSpan(this, 0, length, Spannable.SPAN_INCLUSIVE_INCLUSIVE);
        mMatchesFound = false;
    }

    /**
     * Set the WebView to search.
     *
     * @param webView an implementation of WebView
     */
    public void setWebView(WebView webView) {
        if (null == webView) {
            throw new AssertionError("WebView supplied to "
                    + "FindActionModeCallback cannot be null");
        }
        mWebView = webView;
        mWebView.setFindListener(this);
    }

    @Override
    public void onFindResultReceived(int activeMatchOrdinal, int numberOfMatches,
                                     boolean isDoneCounting) {
        if (isDoneCounting) {
            updateMatchCount(activeMatchOrdinal, numberOfMatches, numberOfMatches == 0);
        }
    }

    /**
     * Move the highlight to the next match.
     * @param next If {@code true}, find the next match further down in the document.
     *             If {@code false}, find the previous match, up in the document.
     */
    private void findNext(boolean next) {
        if (mWebView == null) {
            throw new AssertionError(
                    "No WebView for FindActionModeCallback::findNext");
        }
        if (!mMatchesFound) {
            findAll();
            return;
        }
        if (0 == mNumberOfMatches) {
            // There are no matches, so moving to the next match will not do
            // anything.
            return;
        }
        mWebView.findNext(next);
    }

    /**
     * Highlight all the instances of the string from mEditText in mWebView.
     */
    public void findAll() {
        if (mWebView == null) {
            throw new AssertionError(
                    "No WebView for FindActionModeCallback::findAll");
        }
        CharSequence find = mEditText.getText();
        if (0 == find.length()) {
            mWebView.clearMatches();
            mMatches.setVisibility(View.GONE);
            mMatchesFound = false;
        } else {
            mMatchesFound = true;
            mMatches.setVisibility(View.INVISIBLE);
            mNumberOfMatches = 0;
            mWebView.findAllAsync(find.toString());
        }
    }

    public void showSoftInput() {
        if (mEditText.requestFocus()) {
            mInput.showSoftInput(mEditText, 0);
        }
    }

    public void updateMatchCount(int matchIndex, int matchCount, boolean isEmptyFind) {
        if (!isEmptyFind) {
            mNumberOfMatches = matchCount;
            mActiveMatchIndex = matchIndex;
        } else {
            mMatches.setVisibility(View.GONE);
            mNumberOfMatches = 0;
        }
    }

    // OnClickListener implementation

    @Override
    public void onClick(View v) {
        findNext(true);
    }

    // ActionMode.Callback implementation

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        if (!mode.isUiFocusable()) {
            // If the action mode we're running in is not focusable the user
            // will not be able to type into the find on page field. This
            // should only come up when we're running in a dialog which is
            // already less than ideal; disable the option for now.
            return false;
        }

        mode.setCustomView(mCustomView);
        mode.getMenuInflater().inflate(R.menu.webview_find,
                menu);
        mActionMode = mode;
        Editable edit = mEditText.getText();
        Selection.setSelection(edit, edit.length());
        mMatches.setVisibility(View.GONE);
        mMatchesFound = false;
        mMatches.setText("0");
        mEditText.requestFocus();
        return true;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
        mActionMode = null;
        mWebView.setFindListener(null);
        mInput.hideSoftInputFromWindow(mWebView.getWindowToken(), 0);
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        showSoftInput();

        return false;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        if (mWebView == null) {
            throw new AssertionError(
                    "No WebView for FindActionModeCallback::onActionItemClicked");
        }
        mInput.hideSoftInputFromWindow(mWebView.getWindowToken(), 0);

        int itemId = item.getItemId();
        if (itemId == R.id.find_prev) {
            findNext(false);
        } else if (itemId == R.id.find_next) {
            findNext(true);
        } else {
            return false;
        }
        return true;
    }

    // TextWatcher implementation

    @Override
    public void beforeTextChanged(CharSequence s,
                                  int start,
                                  int count,
                                  int after) {
        // Does nothing.  Needed to implement TextWatcher.
    }

    @Override
    public void onTextChanged(CharSequence s,
                              int start,
                              int before,
                              int count) {
        findAll();
    }

    @Override
    public void afterTextChanged(Editable s) {
        // Does nothing.  Needed to implement TextWatcher.
    }

    private final Rect mGlobalVisibleRect = new Rect();
    private final Point mGlobalVisibleOffset = new Point();
    public int getActionModeGlobalBottom() {
        if (mActionMode == null) {
            return 0;
        }
        View view = (View) mCustomView.getParent();
        if (view == null) {
            view = mCustomView;
        }
        view.getGlobalVisibleRect(mGlobalVisibleRect, mGlobalVisibleOffset);
        return mGlobalVisibleRect.bottom;
    }

    public static class NoAction implements ActionMode.Callback {
        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
        }
    }
}

File: app/src/main/java/app/opendocument/droid/ui/FindActionModeCallback.kt
/*
 * Copyright (C) 2010 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.opendocument.droid.ui

import android.content.Context
import android.text.Editable
import android.text.Selection
import android.text.Spannable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.webkit.WebView
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.view.ActionMode
import app.opendocument.droid.R

// https://github.com/aosp-mirror/platform_frameworks_base/blob/a8b1b1a2e62bcca18f52ed31549c93d43728d152/core/java/android/webkit/FindActionModeCallback.java
class FindActionModeCallback(context: Context) :
    ActionMode.Callback, TextWatcher, View.OnClickListener, WebView.FindListener {

    private val customView: View = LayoutInflater.from(context).inflate(R.layout.webview_find, null)
    private val editText: EditText = customView.findViewById(R.id.edit)
    private val matches: TextView
    private val input: InputMethodManager = context.getSystemService(InputMethodManager::class.java)

    private var webView: WebView? = null
    private var matchesFound = false
    private var numberOfMatches = 0
    private var actionMode: ActionMode? = null

    init {
        editText.setOnClickListener(this)
        setText("")
        matches = customView.findViewById(R.id.matches)
    }

    /**
     * Place text in the text field so it can be searched for. Need to press the find next or find
     * previous button to find all of the matches.
     */
    fun setText(text: String) {
        editText.setText(text)
        val span: Spannable = editText.text
        val length = span.length
        // Ideally, we would like to set the selection to the whole field,
        // but this brings up the Text selection CAB, which dismisses this
        // one.
        Selection.setSelection(span, length, length)
        // Necessary each time we set the text, so that this will watch
        // changes to it.
        span.setSpan(this, 0, length, Spannable.SPAN_INCLUSIVE_INCLUSIVE)
        matchesFound = false
    }

    /**
     * Set the WebView to search.
     *
     * @param webView an implementation of WebView
     */
    fun setWebView(webView: WebView) {
        this.webView = webView
        webView.setFindListener(this)
    }

    private fun requireWebView(caller: String): WebView =
        webView ?: throw AssertionError("No WebView for FindActionModeCallback::$caller")

    override fun onFindResultReceived(
        activeMatchOrdinal: Int,
        numberOfMatches: Int,
        isDoneCounting: Boolean,
    ) {
        if (isDoneCounting) {
            updateMatchCount(activeMatchOrdinal, numberOfMatches, numberOfMatches == 0)
        }
    }

    /**
     * Move the highlight to the next match.
     *
     * @param next If `true`, find the next match further down in the document. If `false`, find the
     *   previous match, up in the document.
     */
    private fun findNext(next: Boolean) {
        val webView = requireWebView("findNext")
        if (!matchesFound) {
            findAll()
            return
        }
        if (0 == numberOfMatches) {
            // There are no matches, so moving to the next match will not do
            // anything.
            return
        }
        webView.findNext(next)
    }

    /** Highlight all the instances of the string from [editText] in the WebView. */
    fun findAll() {
        val webView = requireWebView("findAll")
        val find: CharSequence = editText.text
        if (0 == find.length) {
            webView.clearMatches()
            matches.visibility = View.GONE
            matchesFound = false
        } else {
            matchesFound = true
            matches.visibility = View.INVISIBLE
            numberOfMatches = 0
            webView.findAllAsync(find.toString())
        }
    }

    fun showSoftInput() {
        if (editText.requestFocus()) {
            input.showSoftInput(editText, 0)
        }
    }

    private fun updateMatchCount(matchIndex: Int, matchCount: Int, isEmptyFind: Boolean) {
        if (isEmptyFind) {
            numberOfMatches = 0
            matches.visibility = View.GONE

            return
        }

        numberOfMatches = matchCount
        matches.text = "${matchIndex + 1}/$matchCount"
        matches.visibility = View.VISIBLE
    }

    // OnClickListener implementation

    override fun onClick(v: View) {
        findNext(true)
    }

    // ActionMode.Callback implementation

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        // the AOSP original bailed out here on !mode.isUiFocusable(), which appcompat marks
        // @RestrictTo. we never start this from a dialog.

        mode.customView = customView
        mode.menuInflater.inflate(R.menu.webview_find, menu)
        actionMode = mode
        val edit: Editable = editText.text
        Selection.setSelection(edit, edit.length)
        matches.visibility = View.GONE
        matchesFound = false
        matches.text = "0"
        editText.requestFocus()
        return true
    }

    override fun onDestroyActionMode(mode: ActionMode) {
        actionMode = null

        // tolerant, unlike the paths that need a page to search: the mode is started before
        // setWebView and dismissing it must not throw out of the framework's own callback
        webView?.let {
            it.setFindListener(null)
            input.hideSoftInputFromWindow(it.windowToken, 0)
        }
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        showSoftInput()

        return false
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        val webView = requireWebView("onActionItemClicked")
        input.hideSoftInputFromWindow(webView.windowToken, 0)

        when (item.itemId) {
            R.id.find_prev -> findNext(false)
            R.id.find_next -> findNext(true)
            else -> return false
        }
        return true
    }

    // TextWatcher implementation

    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        // Does nothing.  Needed to implement TextWatcher.
    }

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        findAll()
    }

    override fun afterTextChanged(s: Editable?) {
        // Does nothing.  Needed to implement TextWatcher.
    }
}
